### PR TITLE
schema/cam update schema according to CAM ETSI TS 103 900 V2.1.1

### PR DESCRIPTION
**Pull request by Hugues360**
**Features:**

* Update CAM schema according to CAM ETSI TS 103 900 V2.1.1

Closes: #217

---
**How to test:**

1. Check the [workflow job for CAM schema 2.1.0](https://github.com/Orange-OpenSource/its-client/actions/runs/12320560607)

2. Compare the 2.1.0 schema to the  CAM ETSI TS 103 900 V2.1.1 (https://forge.etsi.org/rep/ITS/asn1/cam_ts103900/-/tree/V2.1.1)

---
**Expected results:**

1. The job succeeds (no validation error occcurs)

2. The 2.1.0 CAM schema should correspond to the Technical Specification

### CODE REVIEW

File: schema/cam/cam_schema_2-1-0.json [hunk 1]
```diff
@@ -126,308 +128,829 @@
...
+                      "type": "integer",
+                      "description": "id of a protected communication zone",
+                      "minimum": 0,
+                      "maximum": 134217727
+                    }
+                  }
+                }
+              }
+            },
+            "rsu_container_high_frequency": {
```
Comment (ymorin-orange) [inline schema/cam/cam_schema_2-1-0.json:524]: `rsu_container_high_frequency` is a new item; it should go in its own commit, rather than be confalted in a commit that fixes the existing schema.

It would also make reviewing easier by splitting the changes in smaller, easier-to-digest changes.

File: schema/cam/cam_schema_2-1-0.json [hunk 2]
```diff
@@ -126,308 +128,829 @@
...
-                    "description": "time travelled by the detecting ITS-S since the previous detected event point (generation_delta_time). tenMilliSecondsInPast(1)",
-                    "minimum": 1,
-                    "maximum": 65535
                   }
                 }
               }
             }
           }
+        },
+        "special_vehicle_container": {
```
Comment (ymorin-orange) [inline schema/cam/cam_schema_2-1-0.json:672]: `special_vehicle_container` is a new item; it should go in its own set of commits, one per new entry in it, rather than be conflated in a commit that fixes the existing schema.

It would also make reviewing easier by splitting the changes in smaller, easier-to-digest changes.

**Review by ymorin-orange — CHANGES_REQUESTED** [review]
Can you please split the big commit into a set of more digestible commits, for example:
- 1 commit to fix the existing schema without adding anything new
- 1 commit adding the RSU container
- 1 commit adding the special_vehicle_container with (e.g.) just public_transport_container
- 1 commit for each of the other special_vehicle_container alternatives...

File: schema/cam/cam_schema_2-1-0.json (hunk 1 above)
Comment (Hugues360) [reply schema/cam/cam_schema_2-1-0.json:524]: Ok, done.

File: schema/cam/cam_schema_2-1-0.json (hunk 2 above)
Comment (Hugues360) [reply schema/cam/cam_schema_2-1-0.json:672]: Ok, done

**Comment by ymorin-orange** [comment]
Thanks for the quick respin! :-) I'll look into it tomorrow morning...

In the meantime, a few comments about the commit log titles (cf process documentation being reviewd internally):

* missing colon in title for commit 2: `schema/cam: add new references field`
* extraneous space after colon in most commit titles
* use imperative form: `s/adding/add/` and `s/updating/update/`
* type: `s/udating/update/`
* warp long lines at ~72 chars

File: schema/cam/cam_schema_2-1-0.json [hunk 3]
```diff
@@ -170,12 +163,28 @@
...
+                  "properties": {
+                    "altitude_value": {
+                      "type": "integer",
+                      "description": "Unit: 0.01 meter. referenceEllipsoidSurface(0), oneCentimeter(1), unavailable(800001)",
+                      "default": 800001,
+                      "minimum": -100000,
+                      "maximum": 800001
+                    },
+                    "altitude_confidence": {
+                      "$comment": "if not provided, 'altitude_confidence' = 15 (unavailable)",
```
Comment (ymorin-orange) [inline schema/cam/cam_schema_2-1-0.json:180]: I wonder where this `$comment` is coming from; it certainly is neither from the TS nor from the EN documents. I also wonder whether it is helpful, and whether it is even correct.

Indeed, `altitude_confidence` is a mandatory property of the `altitude` object, so it can't be "not provided". As such, the comment seems to be contradicting the `required` section for the `altitude` object...

I know there were already such comments in the schema before this change, but no need to add more incorrect such comments...

(Note: this applies to the whole branch, of course, and I'll only comment on this one.)

@mathieu1fb Thoughts?

Comment (ymorin-orange) [reply schema/cam/cam_schema_2-1-0.json:180]: > if a confidence field is mandatory in the spec, but an "unavailable" value exists for it, then we can consider this field optional in the JSON adaptation

I fully support this proposal! It makes so much sense! 👍

Comment (ymorin-orange) [reply schema/cam/cam_schema_2-1-0.json:180]: > if a confidence field is mandatory in the spec, but an "unavailable" value exists for it, then we can consider this field optional in the JSON adaptation

Should we also apply this rule to other properties as well, not just confidence properties?

For example, in the `heading` object, both the `heading_value` and `heading_confidence` properties are required, and they both have an _unavailable_ value. In such a case, I'd say neither properties should be _required_, and thus that the whole `heading` object should also not be _required_ either.

File: schema/cam/cam_schema_2-1-0.json [hunk 4]
```diff
@@ -0,0 +1,958 @@
...
+                "drive_direction",
+                "vehicle_length",
+                "vehicle_width",
+                "longitudinal_acceleration",
+                "curvature",
+                "curvature_calculation_mode",
+                "yaw_rate"
+              ],
+              "properties": {
+                "heading": {
```
Comment (mathieu1fb) [inline schema/cam/cam_schema_2-1-0.json:223]: If we transform each field into objects containing value and confidence (vs having a separate confidence object like before, which was a design choice), I would suggest that each object contains a short `value` and `confidence` field, rather than repeating the object name such as `heading_value` or `heading_confidence`.

File: schema/cam/cam_schema_2-1-0.json (hunk 3 above)
Comment (mathieu1fb) [reply schema/cam/cam_schema_2-1-0.json:180]: I would say that in this example, the `heading` object can then be optional, but if provided, then only its `value` field should be mandatory.

File: schema/cam/cam_schema_2-1-0.json (hunk 4 above)
Comment (ymorin-orange) [reply schema/cam/cam_schema_2-1-0.json:223]: The new layout better matches the layout in the ETSI ASN spec, and so do the names of the properties.

I agree that this new layout is not, err..., very nice and convenient to work with, but I believe it would be diverging too much if we stayed with the previous layout. If ever, that should be done in a major version update where we actually completely depart from the ETSI spec). But for this minor update, I think we should stick to the layout and properties names as defined in the ETSI spec...

**Review by ymorin-orange — CHANGES_REQUESTED** [review]
If you agree with the proposal by @mathieu1fb:
- objects that have a `_value` and a `_confidence` properties, and when both those properties have an `unavalaible` value, then:
    - that object is not `required`
    - the `_value` property for that object is required
    - the `_confidence` property for that object is not required
    - do not add any `$comment` (but keep existing ones to avoid churn)
    - also, it's JSON, so ordering of keys in objects does not matter; where possible, I think the _confidence_ property of an object should be after all the other properties
    - For example, the `altitude` object in `reference_position`:
        ```json
        "reference_position": {
            "type": "object",
            "required": [
                "latitude",
                "longitude"
            ],
            "latitude": {...},
            "longitude": {...},
            "altitude": {
                "type": "object",
                "required": [
                    "altitude_value"
                ],
                "altitude_value": {
                    "type": "integer",
                    "description": "...",
                    "default": 800001,
                    "minimum": -100000,
                    "maximum": 800001
                },
                "altitude_confidence": {
                    "type": "integer",
                    "description": "...",
                    "default": 15,
                    "minimum": 0,
                    "maximum": 15
                }
            },
            "position_confidence_elipse": {...}
        }
        ```

**Comment by Hugues360** [comment]
> If you agree with the proposal by @mathieu1fb:
> 
>     * objects that have a `_value` and a `_confidence` properties, and when both those properties have an `unavalaible` value, then:
>       
>       * that object is not `required`
>       * the `_value` property for that object is required
>       * the `_confidence` property for that object is not required
>       * do not add any `$comment` (but keep existing ones to avoid churn)
>       * also, it's JSON, so ordering of keys in objects does not matter; where possible, I think the _confidence_ property of an object should be after all the other properties
>       * For example, the `altitude` object in `reference_position`:
>         ```json
>         "reference_position": {
>             "type": "object",
>             "required": [
>                 "latitude",
>                 "longitude"
>             ],
>             "latitude": {...},
>             "longitude": {...},
>             "altitude": {
>                 "type": "object",
>                 "required": [
>                     "altitude_value"
>                 ],
>                 "altitude_value": {
>                     "type": "integer",
>                     "description": "...",
>                     "default": 800001,
>                     "minimum": -100000,
>                     "maximum": 800001
>                 },
>                 "altitude_confidence": {
>                     "type": "integer",
>                     "description": "...",
>                     "default": 15,
>                     "minimum": 0,
>                     "maximum": 15
>                 }
>             },
>             "position_confidence_elipse": {...}
>         }
>         ```

Okay. I agree. I am going to update according to this proposal

File: schema/cam/cam_schema_2-1-0.json (hunk 4 above)
Comment (mathieu1fb) [reply schema/cam/cam_schema_2-1-0.json:223]: I agree with following more closely the ETSI message layout, but I still think that simpler `value` / `confidence` pairs for each object would make it clearer and less clutered. CAMs being sent at high rates, it would still be nice to optimize messages a tiny bit (seeing `vehicle_length_value` and `vehicle_length_confidence_indication` scares me).

Comment (ymorin-orange) [reply schema/cam/cam_schema_2-1-0.json:223]: OK, that makes sense. Let's go for shorter names where they make sense.

**Comment by ymorin-orange** [comment]
> * this apply only for objects that have a `_value` and a `_confidence` properties: for example drive_direction remains required even if he has an unavailable value? Idem for all the fields of `position_confidence_ellipse` that will remain required? Or we extend this to all the fields that have a unavailable value?

IMHO, I would say it should also apply to any property that has an `unavailable` value, in fact. @mathieu1fb OK?

> * You asked to leave the existing `$comment`  but as I did not add any , I leave all of them?

Then drop the comments in the objects/properties you move around or otherwise fix. And you did add at least one, in `basic_vehicle_container_high_frequence.performance_class` (im commit "schema/cam: update high frequency container for vehicle"), for example...

**Comment by mathieu1fb** [comment]
> > * this apply only for objects that have a `_value` and a `_confidence` properties: for example drive_direction remains required even if he has an unavailable value? Idem for all the fields of `position_confidence_ellipse` that will remain required? Or we extend this to all the fields that have a unavailable value?
> 
> IMHO, I would say it should also apply to any property that has an `unavailable` value, in fact. @mathieu1fb OK?

I'd say OK, unless it turns out absolutely every field is optional and we cannot make sense of the message that would be sent (let's say a CAM without latitude and longitude...).

**Comment by ymorin-orange** [comment]
> > > * this apply only for objects that have a `_value` and a `_confidence` properties: for example drive_direction remains required even if he has an unavailable value? Idem for all the fields of `position_confidence_ellipse` that will remain required? Or we extend this to all the fields that have a unavailable value?
> > 
> > 
> > IMHO, I would say it should also apply to any property that has an `unavailable` value, in fact. @mathieu1fb OK?
> 
> I'd say OK, unless it turns out absolutely every field is optional and we cannot make sense of the message that would be sent (let's say a CAM without latitude and longitude...).

Yes, there are fields that are really mandatory! Latitude and longitude are, for sure!

> Unless it allows to send lighter CAMs with only what has changed compared to the previous one!

Unlike DENM, I don't think there is a sequence number in a CAM, so we would not know which previous one to update (but a given object may only have a single position anyway. Let's not make that too complex either!)...

File: schema/cam/cam_schema_2-1-0.json [hunk 5]
```diff
@@ -1,6 +1,18 @@
 {
   "$schema": "https://json-schema.org/draft/2020-12/schema",
   "$id": "https://Orange-OpenSource.github.io/its-client/schema/cam",
+  "$references": [
+   {
```
Comment (ymorin-orange) [inline schema/cam/cam_schema_2-1-0.json:5]: Bad indentation; missing one leading space (indentation is 2-space).

File: schema/cam/cam_schema_2-1-0.json [hunk 6]
```diff
@@ -270,78 +349,100 @@
...
-            },
-            "vertical_acceleration": {
-              "type": "integer",
-              "description": "Unit: 0.1 m/s2. pointOneMeterPerSecSquaredUp(1), pointOneMeterPerSecSquaredDown(-1), unavailable(161)",
-              "minimum": -160,
-              "maximum": 161
-            },
-            "confidence": {
-              "$comment": "mandatory on ETSI specification, FIXME?",
+            "steering_wheel_angle":{
```
Comment (ymorin-orange) [inline schema/cam/cam_schema_2-1-0.json]: Missing space before opening curly bracket.

File: schema/cam/cam_schema_2-1-0.json [hunk 7]
```diff
@@ -450,76 +450,79 @@
         },
         "low_frequency_container": {
           "type": "object",
-          "description": "The basic vehicle container low frequency",
-          "required": [
-            "exterior_lights",
-            "path_history"
+          "required":[
```
Comment (ymorin-orange) [inline schema/cam/cam_schema_2-1-0.json:530]: Missing space before the opening square bracket.

File: schema/cam/cam_schema_2-1-0.json [hunk 8]
```diff
@@ -605,6 +605,53 @@
...
+                },
+                "pt_activation": {
+                  "type": "object",
+                  "description": "activation data for real-time systems designed for operations control",
+                  "required": [
+                    "pt_activation_data"
+                  ],
+                  "properties": {
+                    "pt_activation-type": {
+                        "type": "integer",
```
Comment (ymorin-orange) [inline schema/cam/cam_schema_2-1-0.json:668]: Superfluous indentation.

File: schema/cam/cam_schema_2-1-0.json [hunk 9]
```diff
@@ -623,6 +623,11 @@
               "required": [
                 "dangerous_goods_container"
               ]
+             },
```
Comment (ymorin-orange) [inline schema/cam/cam_schema_2-1-0.json:626]: Incorrect indentation.

File: schema/cam/cam_schema_2-1-0.json (hunk 5 above)
Comment (Hugues360) [reply schema/cam/cam_schema_2-1-0.json]: Ok, fixed

File: schema/cam/cam_schema_2-1-0.json (hunk 6 above)
Comment (Hugues360) [reply schema/cam/cam_schema_2-1-0.json]: Ok, fixed

File: schema/cam/cam_schema_2-1-0.json (hunk 7 above)
Comment (Hugues360) [reply schema/cam/cam_schema_2-1-0.json]: Ok, fixed

File: schema/cam/cam_schema_2-1-0.json (hunk 8 above)
Comment (Hugues360) [reply schema/cam/cam_schema_2-1-0.json]: Ok, fixed

File: schema/cam/cam_schema_2-1-0.json (hunk 9 above)
Comment (Hugues360) [reply schema/cam/cam_schema_2-1-0.json]: Ok, fixed.

File: schema/cam/cam_schema_2-1-0.json [hunk 10]
```diff
@@ -633,6 +633,11 @@
               "required": [
                 "rescue_container"
               ]
+            },
+            {
+              "required": [
+                "emergency_container"
```
Comment (ymorin-orange) [inline schema/cam/cam_schema_2-1-0.json:639]: This is a test comment...

File: schema/cam/cam_schema_2-1-0.json [hunk 11]
```diff
@@ -49,7 +49,7 @@
     "version": {
       "type": "string",
       "description": "json message format version",
-      "const": "2.1.0-dev"
+      "const": "2.1.0"
```
Comment (ymorin-orange) [inline schema/cam/cam_schema_2-1-0.json:52]: A second test comment...

File: schema/cam/cam_schema_2-1-0.json [hunk 12]
```diff
@@ -270,78 +349,100 @@
...
+            "performance_class": {
+              "type": "integer",
+              "description": "unavailable 0), performanceClassA(1), performanceClassB(2), reserved(3-7)",
+              "default": 0,
+              "minimum": 0,
+              "maximum": 7
+            },
+            "cen_dsrc_tolling_zone": {
+              "type": "object",
+              "description": "information about a the position of a CEN DRSC Tolling Station operating in the 5,8 GHz frequency band.",
```
Comment (ymorin-orange) [inline schema/cam/cam_schema_2-1-0.json]: Typo: s/DRSC/DSRC/ ([ref](https://forge.etsi.org/rep/ITS/asn1/cam_ts103900/-/blob/V2.1.1/docs/ETSI-ITS-CDD.md#cendsrctollingzone))

**Review by ymorin-orange — APPROVED** [review]
Except for the minor typo, it all seems good to me, now. When you fix that typo and re-push, just merge without requesting another review.

**Comment by Hugues360** [comment]
> Except for the minor typo, it all seems good to me, now. When you fix that typo and re-push, just merge without requesting another review.

Ok and thank you for the review.
